src/QsCompiler/CompilerOptimizations/FunctionEvaluation.fs

```diff
@@ -6,153 +6,147 @@ open Microsoft.Quantum.QsCompiler.DataTypes
         | QsScopeStatement s ->
-            "Not implemented" |> CouldNotEvaluate
+            this.evaluateScope state true s.Body
 
     /// Evaluates a list of Q# statements
-    member private this.evaluateScope (vars: VariablesDict) (enterScope: bool) (scope: QsScope): FunctionState =
-        if enterScope then vars.enterScope()
-        let res =
-            scope.Statements |>
-            Seq.map (this.evaluateStatement vars) |>
-            Seq.tryFind isInterrupt |? Normal
-        if enterScope then vars.exitScope()
-        res
+    member private this.evaluateScope (state: TransformationState) (newScope: bool) (scope: QsScope): FunctionState =
+        result {
+            let stateRef = ref state
+            if newScope then
+                stateRef := enterScope stateRef.Value
+            for stmt in scope.Statements do
+                let! s = this.evaluateStatement stateRef.Value stmt
+                stateRef := s
+            if newScope then
+                stateRef := exitScope stateRef.Value
+            return stateRef.Value
+        }
 
     /// Evaluates a Q# function
-    member this.evaluateFunction (name: QsQualifiedName) (arg: TypedExpression) (types: QsNullable<ImmutableArray<ResolvedType>>): Expr option =
-        let callable = cd.getCallable name
+    member internal this.evaluateFunction (state: TransformationState) (name: QsQualifiedName) (arg: TypedExpression) (types: QsNullable<ImmutableArray<ResolvedType>>): Expr option =
+        let callable = getCallable state name
         QsCompilerError.Verify (
             callable.Specializations.Length = 1,
             "Functions should only have one specialization")
         let impl = callable.Specializations.[0].Implementation
         match impl with
         | Provided (specArgs, scope) ->
-            let vars = VariablesDict()
-            vars.enterScope()
-            fillVars vars (StringTuple.fromQsTuple callable.ArgumentTuple, arg.Expression)
-            let res =
-                match this.evaluateScope vars true scope with
-                | Normal ->
-                    // printfn "Function %O didn't return anything" name.Name.Value
-                    None
-                | Returned expr ->
-                    // printfn "Function %O returned %O" name.Name.Value (prettyPrint expr)
-                    Some expr
-                | Failed expr ->
-                    // printfn "Function %O failed with %O" name.Name.Value (prettyPrint expr)
-                    None
-                | CouldNotEvaluate reason ->
-                    // printfn "Could not evaluate function %O on arg %O for reason %O" name.Name.Value (prettyPrint arg.Expression) reason
-                    None
-            vars.exitScope()
-            res
+            let state = enterScope state
+            let state = defineVarTuple state (toSymbolTuple callable.ArgumentTuple, arg.Expression)
```

Comment:
Here, too, you seem to be missing the specArgs?

Reply:
In the case of functions, is it safe to assume that specArgs is the same as callable.ArgumentTuple?

Reply:
Hm, you should rather not. In the past we have discussed a couple of scenarios where this may not be the case.

Reply:
Fixed in future PR